### PR TITLE
Rewrite RewriteTensorPointer pass to support 2D block store

### PR DESCRIPTION
**Pull request by mfrancepillois**
Prevents `StoreOp`s from being rewritten if it is not suitable for 2D block store.

Addresses Issue: #1143

### CODE REVIEW

**Comment by mfrancepillois** [comment]
Note that the forwarding of the data encoding to the `StoreOp` and the the actual lowering to the 2d block store intrinsic will come in separate PRs.

File: third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp [hunk 1]
```diff
@@ -138,6 +138,12 @@ bool hasDotDpasEncoding(RankedTensorType tensorType) {
   return isa<ttgi::DpasEncodingAttr>(dotLayout.getParent());
 }
 
+bool hasDpasEncoding(RankedTensorType tensorType) {
+  if (auto Enc = tensorType.getEncoding())
```
Comment (FMarno) [inline third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp:142]: I don't think `Enc` follows the style, maybe `enc` or better yet imo `encoding`.

File: third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp [hunk 2]
```diff
@@ -54,6 +54,7 @@ bool isDivisible(Value value, unsigned divisor) {
 /// removed if:
 ///   - the device architecture is not PVC
 ///   - the tensor pointer does not have DotEncoding with DpasEncoding parent
+///   and not have a DpasEncoding
```
Comment (FMarno) [inline third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp:57]:
```suggestion
///   and does not have DpasEncoding
```

File: third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp [hunk 3]
```diff
@@ -729,10 +731,15 @@ class TritonIntelGPURewriteTensorPointerPass
...
         auto src = op->getOperand(0);
-        if (tt::isTensorPointerType(src.getType()))
-          valueToRemove.insert(src);
+        if (tt::isTensorPointerType(src.getType())) {
+          if (auto makeTensorPtrOp = src.getDefiningOp<tt::MakeTensorPtrOp>()) {
+            if (shouldRemove(makeTensorPtrOp, arch))
+              valueToRemove.insert(src);
+          } else {
+            valueToRemove.insert(src);
+          }
```
Comment (FMarno) [inline third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp:741]:
```suggestion
          auto makeTensorPtrOp = src.getDefiningOp<tt::MakeTensorPtrOp>();
          if (!makeTensorPtrOp || shouldRemove(makeTensorPtrOp, arch)) {
            valueToRemove.insert(src);
          }
```

**Review by FMarno — CHANGES_REQUESTED** [review]
I think it would be nice to take some inspiration from the changes to rewrite-tensor-pointer.mlir in this PR https://github.com/intel/intel-xpu-backend-for-triton/pull/1187

File: test/TritonIntelGPU/rewrite-tensor-pointer.mlir [hunk 4]
```diff
@@ -1,10 +1,78 @@
 // RUN: triton-opt %s -split-input-file -tritonintelgpu-rewrite-tensor-pointer | FileCheck %s
 
+// COM: Case 0:
+// COM: Check that operations using block pointers satisfying the following conditions are not rewritten:
+// COM: - the block pointer has the "dot" layout attribute (with dpas parent layout) or has a dpas layout (for store op)
+// COM: - the block pointers is advanced in row major order: strides[order[0]] == 1
```
Comment (chengjunlu) [inline test/TritonIntelGPU/rewrite-tensor-pointer.mlir:6]: From the experience, the row major order is just `strides[1]==1`, which for 2D case, for general case should be `strides[-1]==1`.

Check the use case here of the `order`.
https://github.com/intel/intel-xpu-backend-for-triton/blob/da631c7b442151182ff4f395193e75df15dcc2ea/python/triton/ops/flash_attention.py#L47

File: test/TritonIntelGPU/rewrite-tensor-pointer.mlir [hunk 5]
```diff
@@ -1,10 +1,78 @@
...
+// COM: - the block pointer pitch is divisible by QW: strides[order[1]] % (64 / elemTypeBitWidth) == 0
```
Comment (chengjunlu) [inline test/TritonIntelGPU/rewrite-tensor-pointer.mlir:7]: The row stride should be `strides[0]` which for 2D case, for general case should be `strides[-2]`.

File: third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp (hunk 1 above)
Comment (mfrancepillois) [reply third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp]: The code has been updated this way.

File: third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp (hunk 2 above)
Comment (mfrancepillois) [reply third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp]: I corrected it. Thanks

File: third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp (hunk 3 above)
Comment (mfrancepillois) [reply third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp]: Nice suggestion. The code has been updated this way.

File: test/TritonIntelGPU/rewrite-tensor-pointer.mlir (hunk 4 above)
Comment (mfrancepillois) [reply test/TritonIntelGPU/rewrite-tensor-pointer.mlir]: You're right. The code only checks the contiguous property for strides[1]. I just copied and pasted the comment from `Case 1`, without paying attention. The comment has now been updated this way.

File: test/TritonIntelGPU/rewrite-tensor-pointer.mlir (hunk 5 above)
Comment (mfrancepillois) [reply test/TritonIntelGPU/rewrite-tensor-pointer.mlir]: Same as above. The comment has been updated this way.

File: third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp [hunk 6]
```diff
@@ -729,10 +731,12 @@ class TritonIntelGPURewriteTensorPointerPass
...
       } else if (llvm::isa<tt::StoreOp>(op)) {
-        // TODO: Block store should not be removed when 2d store is enabled
         auto src = op->getOperand(0);
-        if (tt::isTensorPointerType(src.getType()))
-          valueToRemove.insert(src);
+        if (tt::isTensorPointerType(src.getType())) {
+          auto makeTensorPtrOp = src.getDefiningOp<tt::MakeTensorPtrOp>();
+          if (!makeTensorPtrOp || shouldRemove(makeTensorPtrOp, arch))
+            valueToRemove.insert(src);
+        }
```
Comment (LiyangLingIntel) [inline third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp:739]: We can remove the `tt::StoreOp` condition and add it to `LoadOp` case,
```cpp
} else if (llvm::isa<tt::AdvanceOp, tt::LoadOp, tt::StoreOp>(op)) {
        markTensorPointerForRemoval(op->getOperand(0));
} 
```

**Review by FMarno — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp [hunk 7]
```diff
@@ -138,6 +138,12 @@ bool hasDotDpasEncoding(RankedTensorType tensorType) {
   return isa<ttgi::DpasEncodingAttr>(dotLayout.getParent());
 }
 
+bool hasDpasEncoding(RankedTensorType tensorType) {
+  if (auto encoding = tensorType.getEncoding())
+    return isa<ttgi::DpasEncodingAttr>(encoding);
+  return false;
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUTransforms/Utility.cpp:144]:
```suggestion
  return isa_and_nonnull<ttgi::DpasEncodingAttr>(tensorType.getEncoding());
```

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp [hunk 8]
```diff
@@ -717,19 +718,32 @@ class TritonIntelGPURewriteTensorPointerPass
...
-        tt::MakeTensorPtrOp makeTensorPtrOp = getMakeTensorPtrOp(val);
-        if (shouldRemove(makeTensorPtrOp, arch))
-          valueToRemove.insert(val);
+    auto usedByStoreOp = [](Value val) {
+      for (Operation *user : val.getUsers()) {
+        if (llvm::isa<tt::StoreOp>(user))
+          return true;
       }
+      return false;
     };
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUTransforms/RewriteTensorPointer.cpp:727]: This is a bit more readable.
[Suggestion]: 
```
auto usedByStoreOp = [](Value val) { 
   return llvm::any_of(val.getUsers(), [](Operation *user){ return isa<tt.StoreOp>(user); });
};

**Review by etiotto — APPROVED** [review]

**Pull request merged**